### PR TITLE
Don't leak glib error in mono_string_new_checked

### PR DESCRIPTION
**Pull request by lambdageek**
Also added some unit tests for `mono_string_new_checked`

### CODE REVIEW

**Comment by lambdageek** [comment]
@monojenkins build AArch64

File: mono/unit-tests/test-mono-string.c [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+static int
+new_string_ok (void)
+{
+	MonoError error;
+	MonoString *s = mono_string_new_checked (mono_domain_get (), "abcd", &error);
+	mono_error_assert_ok (&error);
+	gunichar2* c = mono_string_chars (s);
+
+	g_assert ((c[0] == (gunichar2)'a') &&
```
Comment (jaykrell) [inline mono/unit-tests/test-mono-string.c]: No need for the casts. This doesn't work on host=ebcdic systems which we probably don't care about. I violate that elsewhere also.

File: mono/unit-tests/test-mono-string.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		  (c[4] == (gunichar2)0));
+	return 0;
+}
+
+static int
+new_string_utf8 (void)
+{
+	MonoError error;
+	const gunichar2 snowman = 0x2603;
+	const char bytes[] = { 0xE2, 0x98, 0x83, 0x00 }; /* U+2603 NUL */
```
Comment (jaykrell) [inline mono/unit-tests/test-mono-string.c]: Can also be "\xE2\x98\x83" but same thing. Also should be const static to demonstrate good form, even if it is only test code. This non-static local actually gets room on the stack and copied to it.

File: mono/unit-tests/test-mono-string.c [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	MonoString *s = mono_string_new_checked (mono_domain_get (), bytes, &error);
+	mono_error_assert_ok (&error);
+	gunichar2* c = mono_string_chars (s);
+	g_assert ((c[0] == snowman) &&
+		  (c[1] == (gunichar2) 0));
```
Comment (jaykrell) [inline mono/unit-tests/test-mono-string.c]: No need for the cast.

File: mono/metadata/object.c [hunk 4]
```diff
@@ -6325,8 +6325,10 @@ mono_string_new_checked (MonoDomain *domain, const char *text, MonoError *error)
 	
 	if (!eg_error)
```
Comment (jaykrell) [inline mono/metadata/object.c]: Great to see this PR in general, so I know where to test lots of string stuff, thank you.

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64

**Pull request merged**